File: baml_language/sdk_tests/crates/python_pydantic2/type_shapes/customizable/roundtrip_tests/test_handles.py
"""Coverage for handle-backed stdlib types returned from BAML to Python.

The non-media cases are encode-back tests: Python receives a generated
class instance with an embedded handle, calls generated stdlib methods
with that same instance, and the engine must see the original handle
state. No external dependency: the HTTP test binds an ephemeral localhost
server and the FS test uses a temp file.
"""

import os
import tempfile
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer

import pytest

import baml_sdk  # noqa: F401  — initializes the BAML runtime
from baml_sdk.baml.fs import open as baml_open
from baml_sdk.baml.http import fetch
from baml_sdk.baml.media import Image

# 1x1 transparent PNG.
PNG_B64 = (
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk"
    "+M8AAAQEAQB9eIv5AAAAAElFTkSuQmCC"
)


# --- media: Image.from_base64 ---------------------------------------------


def test_handles_image_from_base64_roundtrips_payload():
    img = Image.from_base64(PNG_B64, "image/png")
    assert img.mime_type() == "image/png"
    assert img.base64() == PNG_B64


# --- baml.http.Response ---------------------------------------------------

_HTTP_BODY = b"hello from localhost"


class _Handler(BaseHTTPRequestHandler):
    def do_GET(self):  # noqa: N802
        self.send_response(200)
        self.send_header("Content-Type", "text/plain")
        self.send_header("Content-Length", str(len(_HTTP_BODY)))
        self.end_headers()
        self.wfile.write(_HTTP_BODY)

    def log_message(self, *args):  # silence per-request stderr logging
        pass


@pytest.fixture
def http_server():
    srv = HTTPServer(("127.0.0.1", 0), _Handler)
    thread = threading.Thread(target=srv.serve_forever, daemon=True)
    thread.start()
    try:
        yield f"http://127.0.0.1:{srv.server_address[1]}/"
    finally:
        srv.shutdown()
        thread.join()
        srv.server_close()


def test_handles_http_get_response_fields_and_methods(http_server):
    resp = fetch(http_server)
    assert resp.status_code == 200
    assert resp.ok() is True
    assert resp.text() == _HTTP_BODY.decode()


# --- baml.fs.File: cursor state preserved across calls --------------------


@pytest.fixture
def temp_file():
    d = tempfile.mkdtemp()
    path = os.path.join(d, "digits.txt")
    with open(path, "w") as fh:
        fh.write("0123456789")
    try:
        yield path
    finally:
        os.remove(path)
        os.rmdir(d)


def test_handles_open_file_returns_file_handle(temp_file):
    f = baml_open(temp_file, "r")
    assert type(f).__name__ == "File"
    assert f.close() is None


def test_handles_file_cursor_state_persists_across_calls(temp_file):
    f = baml_open(temp_file, "r")

    # Two successive reads on the *same* handle must advance the cursor —
    # the second read continues where the first stopped. This is the
    # load-bearing assertion: engine-side file state survives across
    # separate host→engine FFI calls.
    assert f.read(3) == "012"
    assert f.read(3) == "345"

    # Seek back to the start and confirm the cursor actually moved.
    assert f.seek_from("start", 0) == 0
    assert f.read(2) == "01"

    # text() reads from the current cursor (now at 2) to EOF.
    assert f.text() == "23456789"

    assert f.close() is None

File: baml_language/sdk_tests/crates/cpp/type_shapes/customizable/tests/test_handles.cc
// Coverage for handle-backed stdlib types returned from BAML to C++.
// Port of type_shapes/customizable/roundtrip_tests/test_handles.py.
//
// The non-media cases are encode-back tests: C++ receives a generated
// class instance with an embedded baml::Handle, calls generated stdlib
// methods with that same instance, and the engine must see the original
// handle state. No external dependency: the HTTP test binds an ephemeral
// localhost server and the FS test uses a temp file.
//
// Deviation from the Python file: the HTTP server test is POSIX-only (raw
// sockets stand in for Python's http.server).
#include <baml_sdk.h>
#include <baml_test.h>

#include <cstdio>
#include <optional>
#include <string>

#ifndef _WIN32
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <atomic>
#include <cstring>
#include <thread>
#endif

namespace {

// Writes a small fixture file and removes it on scope exit.
class TempFile {
 public:
  explicit TempFile(const std::string& contents) {
    path_ = "handles_test_digits.txt";
    std::FILE* f = std::fopen(path_.c_str(), "w");
    BAML_ASSERT(f != nullptr);
    std::fwrite(contents.data(), 1, contents.size(), f);
    std::fclose(f);
  }
  ~TempFile() { std::remove(path_.c_str()); }
  const std::string& path() const { return path_; }

 private:
  std::string path_;
};

}  // namespace

// --- media: Image.from_base64 ----------------------------------------------

// 1x1 transparent PNG.
static const char kPngB64[] =
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk"
    "+M8AAAQEAQB9eIv5AAAAAElFTkSuQmCC";

BAML_TEST(image_from_base64_roundtrips_payload) {
  const baml_sdk::baml::media::Image img =
      baml_sdk::baml::media::Image::from_base64(kPngB64,
                                                std::string("image/png"));
  BAML_ASSERT(img.mime_type() == std::optional<std::string>("image/png"));
  BAML_ASSERT_EQ(img.base64(), std::string(kPngB64));
}

// --- baml.fs.File: cursor state preserved across calls --------------------

BAML_TEST(open_file_returns_file_handle) {
  const TempFile tmp("0123456789");
  const baml_sdk::baml::fs::File f = baml_sdk::baml::fs::open(tmp.path(), "r");
  BAML_ASSERT(!f._handle.empty());
  f.close();
}

BAML_TEST(file_cursor_state_persists_across_calls) {
  const TempFile tmp("0123456789");
  const baml_sdk::baml::fs::File f = baml_sdk::baml::fs::open(tmp.path(), "r");

  // Two successive reads on the *same* handle must advance the cursor --
  // the second read continues where the first stopped. This is the
  // load-bearing assertion: engine-side file state survives across
  // separate host->engine FFI calls.
  BAML_ASSERT_EQ(f.read(3), std::string("012"));
  BAML_ASSERT_EQ(f.read(3), std::string("345"));

  // Seek back to the start and confirm the cursor actually moved.
  BAML_ASSERT_EQ(f.seek_from("start", 0), int64_t{0});
  BAML_ASSERT_EQ(f.read(2), std::string("01"));

  // text() reads from the current cursor (now at 2) to EOF.
  BAML_ASSERT_EQ(f.text(), std::string("23456789"));

  f.close();
}

// --- baml.http.Response ----------------------------------------------------

#ifndef _WIN32

namespace {

const char kHttpBody[] = "hello from localhost";

// Minimal one-shot HTTP server: binds an ephemeral localhost port, serves
// exactly one GET with a fixed 200 response on a background thread.
class OneShotHttpServer {
 public:
  OneShotHttpServer() {
    fd_ = ::socket(AF_INET, SOCK_STREAM, 0);
    BAML_ASSERT(fd_ >= 0);
    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = 0;
    BAML_ASSERT(::bind(fd_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) ==
                0);
    socklen_t len = sizeof(addr);
    BAML_ASSERT(::getsockname(fd_, reinterpret_cast<sockaddr*>(&addr), &len) ==
                0);
    port_ = ntohs(addr.sin_port);
    BAML_ASSERT(::listen(fd_, 1) == 0);
    thread_ = std::thread([this]() { ServeOne(); });
  }

  ~OneShotHttpServer() {
    thread_.join();
    ::close(fd_);
  }

  std::string url() const {
    return "http://127.0.0.1:" + std::to_string(port_) + "/";
  }

 private:
  void ServeOne() {
    const int conn = ::accept(fd_, nullptr, nullptr);
    if (conn < 0) {
      return;
    }
    char buf[4096];
    // Read until the end of the request headers (single small GET).
    std::string request;
    while (request.find("\r\n\r\n") == std::string::npos) {
      const ssize_t n = ::recv(conn, buf, sizeof(buf), 0);
      if (n <= 0) {
        break;
      }
      request.append(buf, static_cast<size_t>(n));
    }
    const std::string body = kHttpBody;
    const std::string response =
        "HTTP/1.1 200 OK\r\n"
        "Content-Type: text/plain\r\n"
        "Content-Length: " +
        std::to_string(body.size()) +
        "\r\n"
        "Connection: close\r\n"
        "\r\n" +
        body;
    (void)::send(conn, response.data(), response.size(), 0);
    ::close(conn);
  }

  int fd_ = -1;
  uint16_t port_ = 0;
  std::thread thread_;
};

}  // namespace

BAML_TEST(http_get_response_fields_and_methods) {
  OneShotHttpServer server;
  const baml_sdk::baml::http::Response resp =
      baml_sdk::baml::http::fetch(server.url());
  BAML_ASSERT_EQ(resp.status_code, int64_t{200});
  BAML_ASSERT(resp.ok() == true);
  BAML_ASSERT_EQ(resp.text(), std::string(kHttpBody));
}

#endif  // !_WIN32
